### PR TITLE
8325883: Move Monitor Deflation reporting out of safepoint cleanup

### PR DESCRIPTION
**Pull request by coleenp**
Moved monitor logging out of the safepoint cleanup code to intervals in the MonitorDeflationThread, and made logging come out at the specified level (Info, Debug or Trace).

Tested with tier1-4.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325883](https://bugs.openjdk.org/browse/JDK-8325883): Move Monitor Deflation reporting out of safepoint cleanup (**Enhancement** - P4)

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [0e3919da](https://git.openjdk.org/jdk/pull/18528/files/0e3919da209e4977f2e1c94b4e7230115f50323b)
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18528/head:pull/18528` \
`$ git checkout pull/18528`

Update a local copy of the PR: \
`$ git checkout pull/18528` \
`$ git pull https://git.openjdk.org/jdk.git pull/18528/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18528`

View PR using the GUI difftool: \
`$ git pr show -t 18528`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18528.diff">https://git.openjdk.org/jdk/pull/18528.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18528#issuecomment-2024159845)

### CODE REVIEW

File: src/hotspot/share/runtime/synchronizer.cpp [hunk 1]
```diff
@@ -1950,38 +1950,20 @@ void ObjectSynchronizer::do_final_audit_and_print_stats() {
...
-    audit_and_print_stats(true /* on_exit */);
+    LogStreamHandle(Info, monitorinflation) ls;
+    audit_and_print_stats(&ls, true /* on_exit */);
   }
 }
 
-// This function can be called at a safepoint or it can be called when
-// we are trying to exit the VM. When we are trying to exit the VM, the
-// list walker functions can run in parallel with the other list
-// operations so spin-locking is used for safety.
```
Comment (fisk) [inline src/hotspot/share/runtime/synchronizer.cpp:1962]: Maybe get rid of the spin locking comments. Otherwise this looks great!

**Review by fisk — APPROVED** [review]
Looks good, just need to fix a comment before pushing. Don't need to see the new version.

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/runtime/synchronizer.cpp:1962]: I updated this comment to remove the mention of spin-locking which I think it used to do.  Thanks!

File: src/hotspot/share/runtime/synchronizer.cpp [hunk 2]
```diff
@@ -1950,38 +1950,18 @@ void ObjectSynchronizer::do_final_audit_and_print_stats() {
...
-//
+// This function can be called by the monitor deflation thread or it can be called when
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/synchronizer.cpp:1958]: nit: s/monitor deflation thread/MonitorDeflationThread/

File: src/hotspot/share/runtime/synchronizer.cpp [hunk 3]
```diff
@@ -1993,12 +1973,14 @@ void ObjectSynchronizer::audit_and_print_stats(bool on_exit) {
...
 
-  if ((on_exit && log_is_enabled(Info, monitorinflation)) ||
-      (!on_exit && log_is_enabled(Trace, monitorinflation))) {
-    // When exiting this log output is at the Info level. When called
-    // at a safepoint, this log output is at the Trace level since
-    // there can be a lot of it.
-    log_in_use_monitor_details(ls, !on_exit /* log_all */);
+  // When exiting, only log the interesting entries at the Info level.
+  // When called at intervals by the MonitorDeflationThread, log output at the Trace level since
+  // there can be a lot of it.
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/synchronizer.cpp:1978]: nit: Might want to reflow this comment after "log output" and move the
rest to the next line.

File: test/hotspot/jtreg/runtime/Monitor/MonitorUsedDeflationThresholdTest.java [hunk 4]
```diff
@@ -70,57 +70,99 @@ public static void usage() {
...
-            }
-            System.out.println("too_many='" + too_many + "'");
-            // Uncomment the following line for dumping test output in passing runs:
-            // output_detail.reportDiagnosticSummary();
-
-            System.out.println("PASSED.");
+            // Without args we invoke the tests in a java sub-process.
+            testProcess1();
+            testProcess2();
+            testProcess3();
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/runtime/Monitor/MonitorUsedDeflationThresholdTest.java:165]: Thanks for expanding the testing!

File: test/hotspot/jtreg/runtime/Monitor/MonitorUsedDeflationThresholdTest.java [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021, 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/runtime/Monitor/MonitorUsedDeflationThresholdTest.java:2]: Reviewed this file with whitespace changes disabled. Hopefully I
didn't miss anything that way...

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 1 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/synchronizer.cpp:1962]: Yes, used to have a much more complicated async deflation mechanism and we had to
use spin-locking for safety. @fisk simplified this a long time ago...

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up. I think I only have nit comments here.

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/runtime/synchronizer.cpp:1958]: Fixed, yes, that's better for pattern matching.

File: src/hotspot/share/runtime/synchronizer.cpp (hunk 3 above)
Comment (coleenp) [reply src/hotspot/share/runtime/synchronizer.cpp:1978]: fixed.

File: test/hotspot/jtreg/runtime/Monitor/MonitorUsedDeflationThresholdTest.java (hunk 5 above)
Comment (coleenp) [reply test/hotspot/jtreg/runtime/Monitor/MonitorUsedDeflationThresholdTest.java:2]: My editor strips trailing blanks and my commit command updates copyrights, so you should have seen any other weird formatting in this diff.

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up.

**Comment by coleenp** [comment]
Thanks for reviewing Dan and Erik!

**Pull request closed without merging**